Merge IterationType and EvaluationType, simplify Method interface

vladimir-ch opened the pull request:
A proposal for resolving #103 ~~but still work in progress~~.

* Merged `IterationType` and `EvaluationType` into ~~`RequestType`~~ `Operation`.
* Simplified `Method` interface. It can't be any simpler than this, so if this PR is accepted, further changes are very unlikely.
* The above two modifications ripple through the code and lead at some places to essentially a complete rewrite, most importantly in `LinesearchMethod` and `minimize()`. But the issue at hand is solved, and the code is now much cleaner and more correct. I am ~~quite~~ very happy with the changes (if nobody praises you, you must praise yourself).
* Updated docs, especially for `Method` and `Operation`, although while at it, I updated the docs also at other places, which, as a by-product, resolves issues #29 and #96.

@kortschak @btracey Please take a look.

btracey commented:
I like RequestType, but maybe it's a little too passive? It's not really a request, as the Method will get back what it wants. Perhaps just "Operation"?

The comments around evaluate are good, but the important part to add the information is on Method.

btracey commented:
Reading your other comment now.

vladimir-ch commented:
I like `Operation`, code updated.

I think this is ready for a real review. This PR is quite big, but such a fundamental change cannot be done in small steps. The core changes are in types.go, interfaces.go, local.go and linesearch.go. I will appreciate suggestions for improving the docs. @btracey @kortschak Please take a look.

@btracey Updating NelderMead was easy in the end, the code is well-structured, so it was enough just to add an extra iteration type and a case to a switch. In NelderMead the benefit of this PR is probably most visible. Instead of returning unclear SubIteration and MinorIteration, we simply return FuncEvaluation and when the best value in the simplex could have possibly changed, MajorIteration is returned before proceeding to the reflection step. In both cases the effect is clear even without knowing the context in which the method is used.

btracey commented on local.go:
```diff
@@ -13,12 +13,18 @@ import (
-// Local finds a local minimum of a function using a sequential algorithm.
-// In order to maximize a function, multiply the output by -1.
+// Local finds a local minimum of a minimization problem using a sequential
+// algorithm.
```
I think it's worth saying something along the lines of "A maximization problem can be transformed into a minimization problem by multiplying the function by -1". A lot of 'commercial' solvers allow users to specify either maximize or minimize, so many users new to optimization don't realize the transformation is that easy.

vladimir-ch replied:
Done.

btracey commented on local.go:
```diff
@@ -13,12 +13,18 @@ import (
-// Type switching is used to see if the function implements Gradient, FunctionGradient
-// and Statuser.
+// The first argument represents the problem to be minimized. Its fields are
+// functions that evaluate various quantities related to the problem like the
```
How about it's fields are "routines"? It's obvious from the type that by routine we mean 'go function', and trying to minimize the number of different meanings of the word "function" is probably our own objective function :)

vladimir-ch replied:
Done.

btracey commented on local.go:
```diff
@@ -13,12 +13,18 @@ import (
-// and Statuser.
+// The first argument represents the problem to be minimized. Its fields are
+// functions that evaluate various quantities related to the problem like the
+// objective function or its gradient. p.Func must not be nil and also non-nil
```
How about "that evaluate the objective function, gradient, and other quantities related to the problem". Push the complexity later in the sentence.

btracey replied:
I would say
"The objective function, p.Func, must not be nil. The optimization method used may require other fields to be non-nil as specified by method.Needs(). Local will panic if these are not met. The method can be determined automatically from the supplied method which will be described later."

vladimir-ch replied:
Done.

btracey commented on local.go:
```diff
@@ -124,37 +134,49 @@ func Local(p Problem, initX []float64, settings *Settings, method Method) (*Resu
+
+		case MajorIteration:
+			stats.MajorIterations++
+			if loc.F <= optLoc.F {
```
I don't think we want to have this check any more. The Method is now in control of how OptLoc is maintained, and so if it wants to have a higher F that's it's prerogative. In particular, with this improved control flow, we could add back in the "floating point noise robustness" code.

vladimir-ch replied:
Done.

btracey commented on local.go:
```diff
@@ -124,37 +134,49 @@ func Local(p Problem, initX []float64, settings *Settings, method Method) (*Resu
-		update(loc, optLoc, stats, iterType, startTime)
-		// Get the convergence status before recording the new location.
-		status = checkConvergence(optLoc, iterType, stats, settings)
+		default:
```
Please comment that "default = any of the evaluation cases"

vladimir-ch replied:
Done.

btracey commented on local.go:
```diff
@@ -124,37 +134,49 @@ func Local(p Problem, initX []float64, settings *Settings, method Method) (*Resu
-			status, err = p.Status()
-			if err != nil || status != NotTerminated {
-				return
+		switch op {
```
This is a long for loop. Could you just add a quick comment saying something to the effect of "Sequentially call method.Iterate(), performing the operations it has requested, until convergence"

vladimir-ch replied:
Done.

btracey commented on local.go:
```diff
@@ -246,60 +272,64 @@ func getStartingLocation(p *Problem, method Method, initX []float64, stats *Stat
-		}
-		if loc.F < settings.FunctionThreshold {
-			return FunctionThreshold
+func checkConvergence(loc *Location, settings *Settings) Status {
```
This code is so simple now!

vladimir-ch replied:
Yes! And also evaluate and LinesearchMethod ...

btracey commented on local.go:
```diff
@@ -246,60 +272,64 @@ func getStartingLocation(p *Problem, method Method, initX []float64, stats *Stat
+	return NotTerminated
+}
+
+func checkLimits(loc *Location, stats *Stats, settings *Settings) Status {
```
Please add a comment that this distinguishing feature between checkLimits and checkConvergence is that checkLimits is checked every evaluation, while checkConvergence is only examined at MajorIteration.

vladimir-ch replied:
Done.

btracey commented on local.go:
```diff
@@ -246,60 +272,64 @@ func getStartingLocation(p *Problem, method Method, initX []float64, stats *Stat
 		return FunctionNegativeInfinity
 	}
 
+	if settings.MajorIterations > 0 && stats.MajorIterations >= settings.MajorIterations {
```
Could we update the other checks to use this format? Sorry.

vladimir-ch replied:
Done.

btracey commented on local.go:
```diff
@@ -325,20 +355,18 @@ func checkConvergence(loc *Location, iterType IterationType, stats *Stats, setti
-// detect silent bugs in Methods using inconsistent Location, e.g., using
-// Gradient without a GradEvaluation request.
-func invalidate(loc *Location) {
+// evaluate evaluates the problem p at loc.X using the evaluation eval,
```
evaluate evaluates the routines specified by the Operation at loc.X, storing the answer into loc and updating stats.

vladimir-ch replied:
Done.

btracey commented on types.go:
```diff
@@ -15,54 +15,62 @@ import (
 	HessEvaluation
+
+	// Mask for the evaluating operations.
+	Evaluations = FuncEvaluation | GradEvaluation | HessEvaluation
```
Does this need to be exported?

kortschak replied:
If it is unexported, call it something like evalMask so it's clear in the code what it is doing.

vladimir-ch replied:
Unexported and renamed to evalMask.

btracey commented on types.go:
```diff
@@ -15,54 +15,62 @@ import (
-		e&HessEvaluation != 0,
-		e&^(FuncEvaluation|GradEvaluation|HessEvaluation))
+func (op Operation) IsEvaluation() bool {
+	return op&Evaluations != 0 && op&^Evaluations == 0
```
@kortschak could you confirm you think this line is correct?

kortschak replied:
It is.

btracey commented on types.go:
```diff
@@ -15,54 +15,62 @@ import (
-		e&GradEvaluation != 0,
-		e&HessEvaluation != 0,
-		e&^(FuncEvaluation|GradEvaluation|HessEvaluation))
+func (op Operation) IsEvaluation() bool {
```
Does this need to be exported?

vladimir-ch replied:
Unexported.

btracey commented on local.go:
```diff
@@ -325,20 +355,18 @@ func checkConvergence(loc *Location, iterType IterationType, stats *Stats, setti
+// to NaN. It is the responsibility of Method to assemble a valid Location
+// before requesting MajorIteration.
+func evaluate(p *Problem, loc *Location, eval Operation, stats *Stats) {
+	if !eval.IsEvaluation() {
```
I think this should be in the minimize loop, just after the default case. Check if it's a valid evaluation, and if so evaluate.

vladimir-ch replied:
Done.

btracey commented on types.go:
```diff
@@ -15,54 +15,62 @@ import (
-// information needed at an x location.
-type EvaluationType uint
+// Operation represents the set of operations requested by Method at each
+// iteration. It is a bitmap of *Iteration and *Evaluation constants.
```
That notation is confusing. I originally read it as "Pointer to Iteration type" and "Pointer to evaluation type". 

We need to comment that a Method may return MajorIteration xor NoOperation xor a combination of the evaluations. A method may never return Init and Post. It's probably commented in Method, but I don't think it hurts to have it in multiple places in this case.

vladimir-ch replied:
I thought that using * glob was clever, but I did not realize the connotation with pointer. I removed the * and reorganized the docs a bit.

btracey commented on types.go:
```diff
@@ -15,54 +15,62 @@ import (
-// Evaluation types can be composed together using the binary or operator, for
-// example 'FuncEvaluation | GradEvaluation' to evaluate both the function
-// value and the gradient.
+// Supported Operations. Individual operations must NOT be combined together by
```
Oh, I see. The rules should be in the Operation comment where they will be visible through godoc.

vladimir-ch replied:
I moved the rules to the Operation comment, but I think they would be visible through godoc even if they stayed here.

btracey commented on types.go:
```diff
@@ -15,54 +15,62 @@ import (
 const (
-	NoEvaluation   EvaluationType = 0
-	FuncEvaluation EvaluationType = 1 << (iota - 1)
+	// NoOperation does not specify any operation.
```
NoOperation specifies that no evaluation or iteration should take place.

vladimir-ch replied:
I am not sure about the word "iteration" in this phrase. No sane Method has a reason to return NoOperation from Iterate(), but I can imagine that maybe perhaps it could want to return NoOperation from Init() if it want to just pass the control to Iterate(). At any case, NoOperation will lead to Iterate() being called again without any particular action apart from checking the limits and sending the loc to Recorder.

vladimir-ch replied:
I went with "NoOperation specifies that no evaluation or convergence check should take place."

btracey commented on types.go:
```diff
@@ -15,54 +15,62 @@ import (
-	FuncEvaluation EvaluationType = 1 << (iota - 1)
+	// NoOperation does not specify any operation.
+	NoOperation Operation = 0
+	// InitIteration is sent to Recorder to indicate the initial location. All
```
It's also a signal to methods. "InitIteration is a signal for the initial iteration of the optimization routine. It is sent ..."

kortschak replied:
Put the Operation and EvaluationType constants in separate const blocks so godoc can render them with their type.

vladimir-ch replied:
We don't send Operations to Methods. Previously NelderMead returned InitIteration for the evaluation of the initial simplex, but I think the new framework is cleaner and more consistent. Evaluations of the initial simplex are simply evaluations and when it has been evaluated, MajorIteration is returned because a satisfactory minimum might have been found. So there is no place for returning InitIteration either.

kortschak replied:
Sometimes I don't understand the things that I write. I'll put this one down to temporary colour blindness.

vladimir-ch replied:
Ah, you noticed @kortschak . I kept a tactful silence :-)

kortschak replied:
I noticed that I say foolish things long ago.

You can see that in my github comment history going a fair way back.

btracey commented on types.go:
```diff
@@ -15,54 +15,62 @@ import (
+	// valid.
+	// Methods must not return it.
+	PostIteration
+	// FuncEvaluation is the request to evaluate the objective function.
```
Again, I don't think request is the right word.

vladimir-ch replied:
I changed it to "FuncEvaluation specifies that the objective function should be evaluated." I think using "should" is ok, because Problem.Status can return bad status, so the evaluation might not take place or Iterate() might not be called again.

btracey commented on types.go:
```diff
@@ -111,14 +119,19 @@ type Problem struct {
 	// Func evaluates the objective function at the given location. Func
 	// must not modify x.
 	Func func(x []float64) float64
+
```
Not to change in this PR, but I wonder if we really want it to be the rule that x isn't modified, and instead just pass a copy of x. The copy should be reasonably cheap and then there are no bugs.

vladimir-ch replied:
Good idea. There is always someone who wants to modify x with a good, but misguided intention.

btracey commented on types.go:
```diff
@@ -111,14 +119,19 @@ type Problem struct {
 	// Grad evaluates the gradient at x and stores the result in-place in grad.
 	// Grad must not modify x.
 	Grad func(x []float64, grad []float64)
+
```
Also not for this PR, but I think we should change the input order. The gonum rule is that "answer receivers" are the first inputs.

vladimir-ch replied:
Also agreed.

btracey commented on interfaces.go:
```diff
@@ -5,14 +5,31 @@
 // A Method can optimize an objective function.
+//
+// It uses a reverse-communication interface design represented by the Init()
+// and Iterate() methods to pass and retrieve information from the client. Its
```
"methods to receive information about the objective function". It doesn't pass information to the client aside from X location.

vladimir-ch replied:
It does when MajorIteration is returned. In that case the whole Location might be filled by Method.

I am going to rewrite this paragraph anyway.

btracey commented on interfaces.go:
```diff
@@ -5,14 +5,31 @@
+//
+// It uses a reverse-communication interface design represented by the Init()
+// and Iterate() methods to pass and retrieve information from the client. Its
+// aim is to isolate the method from user-supplied information and data and
```
How about just "This enables automation of common operations like ...". I'm not sure the "isolate" part of the comment is helpful.

vladimir-ch replied:
I changed it to:

```
// This provides independence of the optimization algorithm on user-supplied
// data and their representation, and enables automation of common operations
// like checking for (various types of) convergence and maintaining statistics.
```

I wanted to keep the mention of user data and their representation. With Local() we make and fix all the choices for the user, but there is still the freedom to take a Method and use it in a custom driver routine with different representation of the problem, settings, convergence check, anything.

btracey commented on interfaces.go:
```diff
@@ -5,14 +5,31 @@
+// their representation, and from common operations like checking for (various
+// types of) convergence and maintaining statistics.
+//
+// When Method needs the client to perform an operation like function
```
I would say something like "A Method can command an Evaluation or a MajorIteration operation. An evaluation operation is one or more of the Evaluaton operations (FunctionEvaluation, GradientEvaluation, etc.) which can be combined with the bitwise or operator. In an evaluation operation, the requested routines will be evaluated at the value specified in Location.X. The corresponding fields of Location will be filled with the results from the routine. Alternately, a method can declare a MajorIteration. In a MajorIteration, the values in Location are interpreted as a new minimum. Convergence of the optimization (GradientThreshold, etc.) will be checked using this new minimum."

kortschak replied:
s/Alternately/Alternatively/

Semantically different; alternate is to swap between two state successively, alternative is to choose another - except in NA.

vladimir-ch replied:
I went with your version.

btracey commented on interfaces.go:
```diff
@@ -5,14 +5,31 @@
+// Location upon the next call to Iterate().
+//
+// Methods must not return InitIteration and PostIteration operations. These
+// are reserved for the clients to be passed to Recorders.
```
A method must also not combine the Evaluation operations with the iteration operations.

vladimir-ch replied:
Done.

btracey commented on interfaces.go:
```diff
@@ -5,14 +5,31 @@
-	// Init initializes the method and stores the first location to evaluate
-	// in xNext.
-	Init(loc *Location, xNext []float64) (EvaluationType, IterationType, error)
+	// Init initializes the method based on the initial data in loc, updates it
```
updates the X location in Loc.

vladimir-ch replied:
Like I wrote above, Method can set also other fields of Loc when it is asking for MajorIteration. That's why I didn't write just X.

vladimir-ch replied:
I kept the comment as it is for the moment.

btracey commented on interfaces.go:
```diff
@@ -5,14 +5,31 @@
-	// location to evaluate in xNext.
-	Iterate(loc *Location, xNext []float64) (EvaluationType, IterationType, error)
+	// Iterate retrieves data from loc, performs one iteration of the method,
+	// updates loc and returns the next operation.
```
the x location of loc.

btracey commented:
Did a big chuck of it. Haven't gotten around to the new Linesearcher code or any of the tweaks to the actual Methods.

vladimir-ch commented on local.go:
```diff
@@ -124,37 +134,49 @@ func Local(p Problem, initX []float64, settings *Settings, method Method) (*Resu
 
 		if settings.Recorder != nil {
-			err = settings.Recorder.Record(loc, evalType, iterType, stats)
+			stats.Runtime = time.Since(startTime)
```
I wonder if we shouldn't update Runtime only once per iteration so that the Recorder and checkLimits() get the same value. The complete opposite would be replacing Runtime with StartTime so that the Recorder and checkLimits() are forced to compute the actual runtime at the moment when they need it.

kortschak commented on linesearch.go:
```diff
@@ -10,24 +10,28 @@ import (
-	evalType EvaluationType
-	iterType IterationType
+	first     bool // Indicator of the first iteration.
+	nextMajor bool // Indicates that MajorIteration must be requested at the next call to Iterate().
```
Drop parens.

kortschak commented on interfaces.go:
```diff
@@ -5,14 +5,31 @@
+	// Iterate retrieves data from loc, performs one iteration of the method,
+	// updates loc and returns the next operation.
+	// TODO(vladimir-ch): When decided, say something whether the contents of
+	// Location is preserved between calls to Iterate().
```
Drop parens.

kortschak commented on interfaces.go:
```diff
@@ -5,14 +5,31 @@
-	Init(loc *Location, xNext []float64) (EvaluationType, IterationType, error)
+	// Init initializes the method based on the initial data in loc, updates it
+	// and returns the first operation to be carried out by the caller.
+	// The initial location must be valid as specified by Needs().
```
Drop parens.

kortschak commented on interfaces.go:
```diff
@@ -5,14 +5,31 @@
+// When Method needs the client to perform an operation like function
+// evaluation or check for convergence, it fills the Location with data and
+// returns a corresponding Operation. The result then can be retrieved from
+// Location upon the next call to Iterate().
```
Drop parens.

kortschak commented on interfaces.go:
```diff
@@ -5,14 +5,31 @@
 
 // A Method can optimize an objective function.
+//
+// It uses a reverse-communication interface design represented by the Init()
```
Drop parens.

kortschak commented on interfaces.go:
```diff
@@ -24,6 +41,13 @@ type Method interface {
 
+// Statuser can report the status and any error. It is intended for methods as
+// an additional error reporting mechanism apart from the errors returned from
+// Init() and Iterate().
```
Drop parens.

kortschak commented on linesearch.go:
```diff
@@ -10,24 +10,28 @@ import (
+	loc  Location  // Storage for intermediate locations.
+	eval Operation // Indicator of valid fields in loc.
+
+	lastStep float64   // Step taken from x in the previous call to Iterate().
```
Parens.

kortschak commented on linesearch.go:
```diff
@@ -35,96 +39,149 @@ func (ls *LinesearchMethod) Init(loc *Location, xNext []float64) (EvaluationType
+	switch ls.lastOp {
+	case NoOperation:
+		// TODO(vladimir-ch): We have previously returned with an error and
+		// Init() was not called. What to do? What about ls's internal state?
```
Parens.

btracey replied:
Init() sets last op to NoOp, so we can't tell if init was called or not based on this state.

vladimir-ch replied:
`initNextLinesearch()` then overwrites `lastOp` with the operation returned from `Linesearcher` (which must not return `NoOperation`) so theoretically we could, but probably we should not. We will still have to decide what is the proper thing to do after an error return. I will modify the TODO text and file an issue so that we don't forget.

btracey replied:
Yes, please. The tests are passing now, and we'll be changing anyway with
the copying issue. This way we can think of the algorithm on the whole
without am the extra changes
On Sep 23, 2015 6:16 PM, "Vladimír Chalupecký" <notifications@github.com>
wrote:

> In linesearch.go
> <https://github.com/gonum/optimize/pull/123#discussion_r40274267>:
>
> >  }
> >
> > -func (ls *LinesearchMethod) Iterate(loc *Location, xNext []float64) (EvaluationType, IterationType, error) {
> > -	if ls.iterType == SubIteration {
> > -		// We needed to evaluate invalid fields of Location. Now we have them
> > -		// and can announce MajorIteration.
> > -		copy(xNext, loc.X)
> > -		ls.evalType = NoEvaluation
> > -		ls.iterType = MajorIteration
> > -		return ls.evalType, ls.iterType, nil
> > -	}
> > +func (ls *LinesearchMethod) Iterate(loc *Location) (Operation, error) {
> > +	switch ls.lastOp {
> > +	case NoOperation:
> > +		// TODO(vladimir-ch): We have previously returned with an error and
> > +		// Init() was not called. What to do? What about ls's internal state?
>
> initNextLinesearch() then overwrites lastOp with the operation returned
> from Linesearcher (which must not return NoOperation) so theoretically we
> could, but probably we should not. We will still have to decide what is the
> proper thing to do after an error return. I will modify the TODO text and
> file an issue so that we don't forget.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/gonum/optimize/pull/123/files#r40274267>.
>

kortschak commented on local.go:
```diff
@@ -27,7 +33,7 @@ import (
 // The third argument contains the settings for the minimization. It is here that
 // gradient tolerance, etc. are specified. The DefaultSettings() function
```
Drop parens.

kortschak commented on local.go:
```diff
@@ -36,18 +42,22 @@ import (
+// designed such that they can be reused in future calls to Local.
+//
+// If method implements Statuser, method.Status() is called before every call
+// to method.Iterate(). If the returned Status is not NotTerminated or the
```
Parens.

kortschak commented on neldermead.go:
```diff
@@ -142,10 +143,9 @@ func (n *NelderMead) Init(loc *Location, xNext []float64) (EvaluationType, Itera
+	loc.X[n.fillIdx] += n.SimplexSize
 	n.lastIter = nmInitialize
-	return FuncEvaluation, InitIteration, nil
+	return FuncEvaluation, nil // InitIteration
```
Flesh out this comment or remove it.

vladimir-ch replied:
Ooops, a leftover. Removed.

kortschak commented:
I'm reviewing from a position of ignorance in most of this, so I focussed on wording and code structure. I should know more, but I don't - it would be nice when the API has stabilised to have some wiki pages outlining use cases (examples are probably likely too long to include, but if you think you can do that, that would be good too).

LGTM FWIW

vladimir-ch commented:
@kortschak Thanks for taking a look! I will postpone dropping the parenthesis to another PR. It occurs also at other places in the package, so I will remove them all in one pass later.

@btracey I have addressed your comments. Mostly I took your suggestions for the documentation literally, it was really helpful, they are great. You should look again at what I wrote for Method, especially the first paragraph, and for Operation. After this the main changes to be reviewed are in LinesearchMethod (major changes) and NelderMead (minor changes). The rest is just noise coming from the merge IterationType+EvaluationType => Operation and from the modified Method interface.

vladimir-ch commented on linesearch.go:
```diff
@@ -10,24 +10,28 @@ import (
+	first     bool // Indicator of the first iteration.
+	nextMajor bool // Indicates that MajorIteration must be requested at the next call to Iterate().
+
+	loc  Location  // Storage for intermediate locations.
```
When you get here, @btracey , this is the key point in the discussion about preserving the contents of Location. If we are ok with saying that LinesearchMethod must receive the non-evaluated fields of Location unchanged, then this field can go away and we save (in some cases many) unnecessary copies of Gradient (and Hessian in case of Newton). Instead, we could build the location gradually in the Location passed to Init and Iterate.

For example conjugate gradient produces badly scaled directions and the line searches rarely finish in one iteration. Every point that does not satisfy Armijo or Wolfe means an extra copy. For expensive functions probably not significant, but still ...

btracey replied:
I'm okay with keeping the values in Linesearch. I would like for it to happen in the next PR as this one is already huge (I understand the reason, but we don't need to make it bigger).

vladimir-ch commented on linesearch.go:
```diff
@@ -35,96 +39,152 @@ func (ls *LinesearchMethod) Init(loc *Location, xNext []float64) (EvaluationType
+			ls.loc.F = loc.F
+		}
+		if ls.lastOp&GradEvaluation != 0 {
+			copy(ls.loc.Gradient, loc.Gradient)
```
This ...

vladimir-ch commented on linesearch.go:
```diff
@@ -35,96 +39,152 @@ func (ls *LinesearchMethod) Init(loc *Location, xNext []float64) (EvaluationType
+			copy(ls.loc.Gradient, loc.Gradient)
+		}
+		if ls.lastOp&HessEvaluation != 0 {
+			ls.loc.Hessian.CopySym(loc.Hessian)
```
... and this are the copies that could go away. Especially this copy will be the most wasteful. Newton will often finish the linesearch with just one iteration, then HessEvaluation will be returned, then here a copy to ls.loc and then few line below a copy of the whole location back to loc.

btracey commented on interfaces.go:
```diff
@@ -5,14 +5,40 @@
+// data and their representation, and enables automation of common operations
+// like checking for (various types of) convergence and maintaining statistics.
+//
+// A Method can command an Evaluation or a MajorIteration operation.
```
We need to add NoIteration

vladimir-ch replied:
Done

btracey commented on linesearch.go:
```diff
@@ -35,96 +39,152 @@ func (ls *LinesearchMethod) Init(loc *Location, xNext []float64) (EvaluationType
+
+	case MajorIteration:
+		// We previously requested MajorIteration but since we're here, the
+		// previous location was not good enough to converge. So start the next
```
Add "converge the full optimization", just so there's no confusion with "converge the line search". Remove the "So" (don't pick that up from me :) )

vladimir-ch replied:
Done. Done (I won't :) ).

btracey commented on linesearch.go:
```diff
@@ -35,96 +39,152 @@ func (ls *LinesearchMethod) Init(loc *Location, xNext []float64) (EvaluationType
-		ls.evalType = NoEvaluation
-		ls.iterType = NoIteration
-		return ls.evalType, ls.iterType, ErrNoProgress
+	if step != ls.lastStep {
```
Would you mind swapping the if to have the equals case first? My first thought is "what happens if they're equal" and want to read below. Plus, that case is simpler and easy to digest.

vladimir-ch replied:
Done

btracey commented:
LGTM I think with the few minor cleanups.

btracey commented:
Thanks! This is a very nice fix to the major problem in the package design.
